providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -513,135 +487,122 @@ static const OSSL_PARAM *mlkem_gen_settable_params(ossl_unused void *genctx,
-            || !ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
-                                                  &mkey->privkey)
-            || !ossl_mlkem768_public_from_private(&mkey->pubkey, &mkey->privkey)
-            || !ossl_mlkem768_marshal_public_key(mkey->encoded_pubkey,
-                                                 &mkey->pubkey))
-            goto err;
-    } else {
-        debug_print("MLKEMKM generate random keys");
-        if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
-                                        &mkey->privkey, mkey->mlkem_ctx)
-            || !ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
-                                                  &mkey->privkey)
-            || !ossl_mlkem768_public_from_private(&mkey->pubkey,
-                                                  &mkey->privkey))
-            goto err;
-    }
-
-    debug_print("MLKEMKM gen returns set %p\n", mkey);
-    return mkey;
-
-err:
-    OPENSSL_free(mkey);
+    ossl_mlkem_vcleanse_prvkey(v, &key->prvkey);
+    OPENSSL_free(key);
     return NULL;
 }
 
-static void mlkem_gen_cleanup(void *genctx)
+static void mlkem_gen_cleanup(void *vgctx)
 {
-    struct mlkem_gen_ctx *gctx = genctx;
-
-    OPENSSL_free(gctx->seed);
-    debug_print("MLKEMKM gen cleanup for %p\n", gctx);
-    OPENSSL_free(gctx);
+    OPENSSL_free(vgctx);
 }
 
-static void *mlkem_dup(const void *vsrckey, int selection)
+static void *mlkem_dup(const void *vkey, int selection)
 {
-    const MLKEM768_KEY *srckey = (const MLKEM768_KEY *)vsrckey;
-    MLKEM768_KEY *dstkey;
+    const MLKEM_KEY *key = (const MLKEM_KEY *)vkey;
+    vinfo_t v = key->vinfo;
+    MLKEM_KEY *newkey;
 
-    debug_print("MLKEMKM dup called for %p\n", srckey);
     if (!ossl_prov_is_running())
         return NULL;
 
-    dstkey = mlkem_new(srckey->provctx);
-    if (dstkey == NULL)
+    if ((newkey = OPENSSL_zalloc(sizeof(*key))) == NULL)
         return NULL;
+    if ((newkey->ctx = ossl_mlkem_ctx_dup(key->ctx)) == NULL)
+        goto err;
+    newkey->provctx = key->provctx;
+    newkey->vinfo = v;
 
-    dstkey->keytype = srckey->keytype;
-    if (srckey->encoded_pubkey != NULL
-            && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
-        memcpy((void *)&dstkey->pubkey, (void *)&srckey->pubkey,
-               sizeof(srckey->pubkey));
-        if ((dstkey->encoded_pubkey = OPENSSL_memdup(srckey->encoded_pubkey,
-                                                     OSSL_MLKEM768_PUBLIC_KEY_BYTES)) == NULL)
-            return NULL;
+    /* If key material is requested, clone the entire structure */
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return newkey;
+
+    if (key->prvkey != NULL
+        && (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
+        newkey->prvkey = OPENSSL_memdup(key->prvkey, v->prvalloc);
+        if (newkey->prvkey != NULL)
+            return newkey;
```

Comment:
Is it impossible that the source key also contains public key material? If it did, it wouldn't be copied over due to this return, right?

Reply:
An embedded public key is an integral part of the private key. This is fine.